File: packages/edge/src/middleware/server-auth.ts
import { Request, Response, NextFunction } from 'express';
import { tokenMirrorService } from '../services/token-mirror-service';
import { environmentRegistry } from '../services/environment-registry';
import { ErrorCodes, sendUnauthorized } from '../utils/api-response';

// Unsecured token format: unsecured-{org}:{project}:{env}-{type}-api-token
const UNSECURED_TOKEN_REGEX =
  /^unsecured-([^:]+):([^:]+):(.+)-(server|client|edge)-api-token$/;

// Legacy unsecured tokens — auto-resolve to default/default/development
const LEGACY_TOKENS: Record<string, boolean> = {
  'unsecured-client-api-token': true,
  'unsecured-server-api-token': true,
  'unsecured-edge-api-token': true,
};
const LEGACY_ENV_NAME = 'development';

/**
 * Extended request with resolved environment
 */
export interface ServerRequest extends Request {
  environmentId?: string;
  cacheKey?: string;
  applicationName?: string;
}

/**
 * Server SDK authentication middleware for Edge
 * Resolves environment from token (no :env path parameter)
 */
export function serverAuth(
  req: ServerRequest,
  res: Response,
  next: NextFunction
): void {
  const apiToken = req.headers['x-api-token'] as string;

  if (!apiToken) {
    sendUnauthorized(
      res,
      'x-api-token header is required',
      ErrorCodes.AUTH_TOKEN_REQUIRED
    );
    return;
  }

  // 1. Try unsecured token format: unsecured-{org}:{project}:{env}-{type}-api-token
  const unsecuredMatch = apiToken.match(UNSECURED_TOKEN_REGEX);
  if (unsecuredMatch) {
    const [, , , envId] = unsecuredMatch;
    // The token IS the cache key for the SDK
    req.cacheKey = apiToken;
    // Resolve actual environment ID from the registry
    req.environmentId =
      environmentRegistry.resolveEnvironmentId(envId) || envId;
    req.applicationName =
      (req.headers['x-application-name'] as string) || 'unknown';
    return next();
  }

  // 2. Legacy unsecured tokens → resolve to development
  if (LEGACY_TOKENS[apiToken]) {
    const cacheKey =
      environmentRegistry.resolveEnvironmentToken(LEGACY_ENV_NAME);
    if (!cacheKey) {
      sendUnauthorized(
        res,
        'Could not resolve environment for legacy token',
        ErrorCodes.AUTH_TOKEN_INVALID
      );
      return;
    }
    req.cacheKey = cacheKey;
    // Get actual environment ID
    req.environmentId =
      environmentRegistry.resolveEnvironmentId(LEGACY_ENV_NAME) ||
      LEGACY_ENV_NAME;
    req.applicationName =
      (req.headers['x-application-name'] as string) || 'unknown';
    return next();
  }

  // 3. Real production token — validate and resolve environment
  const validation = tokenMirrorService.validateToken(apiToken, 'server');
  if (!validation.valid) {
    sendUnauthorized(
      res,
      'Invalid or unauthorized server API token',
      ErrorCodes.AUTH_TOKEN_INVALID
    );
    return;
  }

  // Resolve environment from token (token:environment is 1:1)
  const token = validation.token;
  const tokenEnvId = token?.environmentId;
  if (!tokenEnvId) {
    sendUnauthorized(
      res,
      'Token does not have a specific environment binding',
      ErrorCodes.AUTH_TOKEN_INVALID
    );
    return;
  }

  const cacheKey = environmentRegistry.resolveEnvironmentToken(tokenEnvId);
  if (!cacheKey) {
    sendUnauthorized(
      res,
      `Could not resolve environment: ${tokenEnvId}`,
      ErrorCodes.AUTH_TOKEN_INVALID
    );
    return;
  }

  req.cacheKey = cacheKey;
  req.environmentId =
    environmentRegistry.resolveEnvironmentId(tokenEnvId) || tokenEnvId;
  req.applicationName =
    (req.headers['x-application-name'] as string) || 'unknown';

  next();
}

File: packages/edge-rust/src/middleware/server_auth.rs
// Gatrix Edge Rust - Server Authentication
// Validates server API tokens and resolves environment context

use std::sync::Arc;
use actix_web::{HttpRequest, HttpResponse};
use log::{debug, warn};
use regex::Regex;
use std::sync::LazyLock;

use crate::services::token_mirror::TokenMirrorService;
use crate::services::environment_registry::EnvironmentRegistry;

/// Unsecured token regex
static UNSECURED_TOKEN_REGEX: LazyLock<Regex> = LazyLock::new(|| {
  Regex::new(r"^unsecured-([^:]+):([^:]+):(.+)-(server|client|edge)-api-token$").unwrap()
});

/// Legacy unsecured tokens
const LEGACY_TOKENS: &[&str] = &[
  "unsecured-client-api-token",
  "unsecured-server-api-token",
  "unsecured-edge-api-token",
];

const LEGACY_ENV_NAME: &str = "development";

/// Server context extracted from authentication
#[derive(Debug, Clone)]
pub struct ServerContext {
  pub environment_id: String,
  pub cache_key: String,
  pub application_name: String,
}

/// Authenticate a server request and return ServerContext
pub async fn authenticate_server(
  req: &HttpRequest,
  token_mirror: &Arc<TokenMirrorService>,
  env_registry: &Arc<EnvironmentRegistry>,
) -> Result<ServerContext, HttpResponse> {
  let api_token = req
    .headers()
    .get("x-api-token")
    .and_then(|v| v.to_str().ok())
    .ok_or_else(|| {
      HttpResponse::Unauthorized().json(serde_json::json!({
        "success": false,
        "error": {
          "code": "AUTH_TOKEN_REQUIRED",
          "message": "x-api-token header is required"
        }
      }))
    })?;

  let application_name = req
    .headers()
    .get("x-application-name")
    .and_then(|v| v.to_str().ok())
    .unwrap_or("unknown")
    .to_string();

  // 1. Try unsecured token format
  if let Some(captures) = UNSECURED_TOKEN_REGEX.captures(api_token) {
    let env_id = captures.get(3).unwrap().as_str();
    let environment_id = env_registry
      .resolve_environment_id(env_id)
      .await
      .unwrap_or_else(|| env_id.to_string());

    debug!("Server auth: unsecured token, env={}", environment_id);

    return Ok(ServerContext {
      environment_id: environment_id.clone(),
      cache_key: environment_id,
      application_name,
    });
  }

  // 2. Legacy unsecured tokens
  if LEGACY_TOKENS.contains(&api_token) {
    let env_id = env_registry
      .resolve_environment_id(LEGACY_ENV_NAME)
      .await
      .ok_or_else(|| {
        HttpResponse::Unauthorized().json(serde_json::json!({
          "success": false,
          "error": {
            "code": "AUTH_TOKEN_INVALID",
            "message": "Could not resolve environment for legacy token"
          }
        }))
      })?;

    debug!("Server auth: legacy token, env={}", env_id);

    return Ok(ServerContext {
      environment_id: env_id.clone(),
      cache_key: env_id,
      application_name,
    });
  }

  // 3. Real production token
  let validation = token_mirror.validate_token(api_token, "server").await;
  if !validation.valid {
    warn!("Server authentication failed");
    return Err(HttpResponse::Unauthorized().json(serde_json::json!({
      "success": false,
      "error": {
        "code": "AUTH_TOKEN_INVALID",
        "message": "Invalid or unauthorized server API token"
      }
    })));
  }

  let token = validation.token.as_ref();
  let token_env_id = token
    .and_then(|t| t.environment_id.as_ref())
    .ok_or_else(|| {
      HttpResponse::Unauthorized().json(serde_json::json!({
        "success": false,
        "error": {
          "code": "AUTH_TOKEN_INVALID",
          "message": "Token does not have a specific environment binding"
        }
      }))
    })?;

  let env_id = env_registry
    .resolve_environment_id(token_env_id)
    .await
    .unwrap_or_else(|| token_env_id.clone());

  if !env_registry.has_environment(&env_id).await {
    return Err(HttpResponse::Unauthorized().json(serde_json::json!({
      "success": false,
      "error": {
        "code": "AUTH_TOKEN_INVALID",
        "message": format!("Could not resolve environment: {}", token_env_id)
      }
    })));
  }

  debug!("Server authenticated, env={}", env_id);

  Ok(ServerContext {
    environment_id: env_id.clone(),
    cache_key: env_id,
    application_name,
  })
}
